File: java/src/main/java/dev/nezo/burmaldaholic/games/poker/logic/Bots.java
package dev.nezo.burmaldaholic.games.poker.logic;

import dev.nezo.burmaldaholic.games.poker.logic.Hand.Action;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * House bots: Fish / Regular / Shark (GAME_DESIGN.md §7.4). A bot sees only its own cards and public
 * information (no collusion): the game builds a {@link View} with {@link #view}, computes the equity
 * when {@link #samplesFor} says so, then calls {@link #decide}.
 */
public final class Bots {
	public enum Tier {
		FISH, REGULAR, SHARK;

		public String id() {
			return name().toLowerCase(Locale.ROOT);
		}

		public static Tier byId(String id) {
			for (Tier t : values()) {
				if (t.id().equals(id)) {
					return t;
				}
			}
			return REGULAR;
		}
	}

	public enum Position { EARLY, MIDDLE, LATE, SB, BB }

	/** Fixed, non-localized bot names (GAME_DESIGN.md §7.4, same list as Bedrock). */
	public static final List<String> NAMES = List.of("Lucky Steve", "Grandpa Pavel", "Creeper42", "Mr. Blocksworth", "Diamond Dave",
		"Aunt Zoya", "Redstone Rick", "Nether Nick", "Emerald Emma", "Sir Oinksalot", "Baba Valya", "Enderman Ed");

	/**
	 * What a bot may know.
	 *
	 * @param pot              all chips in the middle (pots + current street bets)
	 * @param limpers          preflop: players who just called the big blind
	 * @param facingRaise      preflop: someone raised above the big blind
	 * @param preflopRaiser    this bot made the last preflop raise
	 * @param opponents        opponents still in the hand
	 * @param loosestHumanVpip highest VPIP (0..1) among live human opponents, -1 if unknown
	 */
	public record View(int[] hole, int[] board, Hand.Street street, long pot, long toCall, long stack, long bet, long currentBet,
			long bb, boolean canCheck, boolean canRaise, long minRaiseTo, long maxRaiseTo, Position position, int limpers,
			boolean facingRaise, boolean preflopRaiser, int opponents, double loosestHumanVpip) {}

	private Bots() {}

	/** Preflop position of player i (early = first 2 to act, late = cutoff/button). */
	public static Position positionOf(Hand h, int i) {
		int n = h.players().size();
		if (n == 2) {
			return i == h.button() ? Position.SB : Position.BB;
		}
		if (i == h.sbIndex()) {
			return Position.SB;
		}
		if (i == h.bbIndex()) {
			return Position.BB;
		}
		if (i == h.button()) {
			return Position.LATE;
		}
		int first = (h.bbIndex() + 1) % n;
		int order = Math.floorMod(i - first, n);
		if (order < 2) {
			return Position.EARLY;
		}
		if ((i + 1) % n == h.button()) {
			return Position.LATE;
		}
		return Position.MIDDLE;
	}

	/** Public view for the bot at hand index i. {@code vpip}: human id → VPIP (0..1). */
	public static View view(Hand h, int i, Map<String, Double> vpip) {
		Hand.Player p = h.player(i);
		Hand.Legal l = h.legal(i);
		List<Integer> live = h.livePlayers();
		double loosest = -1;
		int opponents = 0;
		for (int j : live) {
			if (j == i) {
				continue;
			}
			opponents++;
			Hand.Player q = h.player(j);
			Double v = q.human && vpip != null ? vpip.get(q.id) : null;
			if (v != null) {
				loosest = Math.max(loosest, v);
			}
		}
		int limpers = 0;
		boolean preflop = h.street() == Hand.Street.PREFLOP;
		if (preflop && h.currentBet() <= h.bb()) {
			for (int j = 0; j < h.players().size(); j++) {
				Hand.Player q = h.player(j);
				if (j != i && q.vpip && !q.pfr) {
					limpers++;
				}
			}
		}
		int[] board = new int[h.board().size()];
		for (int k = 0; k < board.length; k++) {
			board[k] = h.board().get(k);
		}
		return new View(p.hole(), board, h.street(), h.potTotal(), l.toCall(), p.stack, p.bet, h.currentBet(), h.bb(), l.canCheck(),
			l.canRaise(), l.minRaiseTo(), l.maxRaiseTo(), positionOf(h, i), limpers, preflop && h.currentBet() > h.bb(),
			h.preflopRaiser() == i, opponents, loosest);
	}

	/** Shark equity ranges: live opponents who raised preflop hold a top-40 % Chen hand. */
	public static int[] opponentRanges(Hand h, int i) {
		List<Integer> out = new ArrayList<>();
		for (int j : h.livePlayers()) {
			if (j != i) {
				out.add(h.player(j).pfr ? Equity.top40Chen() : Integer.MIN_VALUE);
			}
		}
		return out.stream().mapToInt(Integer::intValue).toArray();
	}

	/** Monte-Carlo samples a tier needs on this street (0 = no simulation). */
	public static int samplesFor(Tier tier, Hand.Street street, int regularSamples, int sharkSamples) {
		if (tier == Tier.FISH || street == Hand.Street.PREFLOP) {
			return 0;
		}
		return tier == Tier.REGULAR ? regularSamples : sharkSamples;
	}

	/** Picks a tier from Fish/Regular/Shark percentages (normalized; all zero = Regular). */
	public static Tier pickTier(PokerRng rng, int[] mix) {
		int total = 0;
		int[] w = new int[3];
		for (int k = 0; k < 3; k++) {
			w[k] = mix != null && k < mix.length ? Math.max(0, mix[k]) : 0;
			total += w[k];
		}
		if (total <= 0) {
			return Tier.REGULAR;
		}
		int roll = rng.nextInt(total);
		for (int k = 0; k < 3; k++) {
			roll -= w[k];
			if (roll < 0) {
				return Tier.values()[k];
			}
		}
		return Tier.SHARK;
	}

	/** Total street bet for a bet/raise of {@code fraction} of the pot. */
	public static long sizeTo(View v, double fraction) {
		if (v.currentBet() == 0) {
			return Math.max(v.bb(), Math.round(v.pot() * fraction));
		}
		return Math.round(v.currentBet() + fraction * (v.pot() + v.toCall()));
	}

	private static Action passive(View v) {
		return v.canCheck() ? Action.check() : Action.fold();
	}

	private static Action callOrCheck(View v) {
		return v.toCall() > 0 ? Action.call() : Action.check();
	}

	private static double potOdds(View v) {
		return v.toCall() > 0 ? (double) v.toCall() / (v.pot() + v.toCall()) : 0;
	}

	/** Made-hand category that uses at least one hole card (0 = nothing beyond the board). */
	public static int madeCategory(int[] hole, int[] board) {
		if (board.length < 3) {
			return 0;
		}
		int[] all = new int[2 + board.length];
		all[0] = hole[0];
		all[1] = hole[1];
		System.arraycopy(board, 0, all, 2, board.length);
		int cat = HandEvaluator.category(HandEvaluator.evaluate(all));
		return cat > boardCategory(board) ? cat : 0;
	}

	private static int boardCategory(int[] board) {
		if (board.length >= 5) {
			return HandEvaluator.category(HandEvaluator.evaluate(board, Math.min(board.length, 7)));
		}
		int[] counts = new int[15];
		for (int c : board) {
			counts[Cards.rank(c)]++;
		}
		int first = 0, second = 0;
		for (int n : counts) {
			if (n > first) {
				second = first;
				first = n;
			} else if (n > second) {
				second = n;
			}
		}
		if (first == 4) {
			return HandEvaluator.FOUR_OF_A_KIND;
		}
		if (first == 3) {
			return second == 2 ? HandEvaluator.FULL_HOUSE : HandEvaluator.THREE_OF_A_KIND;
		}
		if (first == 2) {
			return second == 2 ? HandEvaluator.TWO_PAIR : HandEvaluator.PAIR;
		}
		return HandEvaluator.HIGH_CARD;
	}

	private static boolean pocketPair(int[] h) {
		return Cards.rank(h[0]) == Cards.rank(h[1]);
	}

	private static boolean suited(int[] h) {
		return Cards.suit(h[0]) == Cards.suit(h[1]);
	}

	private static boolean connected(int[] h) {
		return Math.abs(Cards.rank(h[0]) - Cards.rank(h[1])) == 1;
	}

	private static boolean hasAce(int[] h) {
		return Cards.rank(h[0]) == 14 || Cards.rank(h[1]) == 14;
	}

	// ---- Fish -------------------------------------------------------------------------------------

	private static Action fish(View v, PokerRng rng) {
		double jitter = 0.9 + rng.nextDouble() * 0.2;
		if (v.street() == Hand.Street.PREFLOP) {
			int c = Equity.chen(v.hole()[0], v.hole()[1]);
			if (c >= 12 && v.currentBet() < 3 * v.bb()) {
				return Action.raiseTo((long) (3 * v.bb() * jitter));
			}
			if (v.facingRaise()) {
				if (pocketPair(v.hole()) && v.currentBet() <= 10 * v.bb()) {
					return callOrCheck(v);
				}
				if (c >= 6 && v.toCall() <= 0.2 * (v.stack() + v.bet())) {
					return callOrCheck(v);
				}
				return passive(v);
			}
			return c >= 4 ? callOrCheck(v) : passive(v);
		}
		int made = madeCategory(v.hole(), v.board());
		if (v.toCall() > 0) {
			return made >= 1 && v.toCall() <= v.pot() ? Action.call() : Action.fold();
		}
		if (made >= 2) {
			return Action.raiseTo(sizeTo(v, 0.5 * jitter));
		}
		if (made == 0 && rng.nextDouble() < 0.05) {
			return Action.raiseTo(sizeTo(v, 0.5 * jitter));
		}
		return Action.check();
	}

	// ---- Regular / Shark preflop ---------------------------------------------------------------------

	private static int openThreshold(Position pos) {
		return switch (pos) {
			case EARLY -> 8;
			case MIDDLE, BB -> 7;
			case LATE, SB -> 6;
		};
	}

	private static Action preflopSolid(View v, PokerRng rng, boolean shark) {
		int c = Equity.chen(v.hole()[0], v.hole()[1]);
		int d = shark ? 1 : 0;
		int widen = shark && v.loosestHumanVpip() > 0.4 ? 1 : 0;
		if (!v.facingRaise()) {
			if (c >= openThreshold(v.position()) - d) {
				return Action.raiseTo(3 * v.bb() + v.limpers() * v.bb());
			}
			return passive(v);
		}
		if (c >= 11 - d) {
			return Action.raiseTo(3 * v.currentBet());
		}
		if (shark && suited(v.hole()) && (connected(v.hole()) || hasAce(v.hole())) && rng.nextDouble() < 0.08) {
			return Action.raiseTo(3 * v.currentBet());
		}
		if (c >= 9 - d - widen) {
			return callOrCheck(v);
		}
		return passive(v);
	}

	// ---- Regular / Shark postflop -------------------------------------------------------------------

	private static Action regularPost(View v, double e, PokerRng rng) {
		if (e >= 0.65) {
			return Action.raiseTo(sizeTo(v, 0.66));
		}
		if (v.toCall() > 0) {
			return e >= potOdds(v) + 0.05 ? Action.call() : Action.fold();
		}
		if (v.street() == Hand.Street.FLOP && v.preflopRaiser() && rng.nextDouble() < 0.3) {
			return Action.raiseTo(sizeTo(v, 0.5));
		}
		return Action.check();
	}

	private static Action sharkPost(View v, double e, PokerRng rng) {
		if (e >= 0.85) {
			double r = rng.nextDouble();
			if (r < 0.2) {
				return Action.allIn();
			}
			if (r < 0.35) {
				return callOrCheck(v);
			}
			return Action.raiseTo(sizeTo(v, 0.75));
		}
		if (e >= 0.6) {
			return v.toCall() > 0 ? Action.call() : Action.raiseTo(sizeTo(v, 0.5 + rng.nextDouble() * 0.25));
		}
		if (v.street() == Hand.Street.FLOP && e >= 0.3 && rng.nextDouble() < 0.4) {
			return Action.raiseTo(sizeTo(v, 0.5 + rng.nextDouble() * 0.25));
		}
		if (v.toCall() > 0) {
			return e >= potOdds(v) ? Action.call() : Action.fold();
		}
		return Action.check();
	}

	/** One step less aggressive: raise → call/check, call → fold/check. */
	private static Action lower(View v, Action a) {
		return switch (a.kind()) {
			case RAISE, ALL_IN -> callOrCheck(v);
			case CALL -> passive(v);
			default -> a;
		};
	}

	/** Makes an action legal for the view (clamps sizes, raise → call when raising is closed). */
	public static Action legalize(View v, Action a) {
		return switch (a.kind()) {
			case RAISE, ALL_IN -> {
				if (!v.canRaise()) {
					yield callOrCheck(v);
				}
				if (a.kind() == Action.Kind.ALL_IN) {
					yield a;
				}
				long to = Math.max(v.minRaiseTo(), Math.min(v.maxRaiseTo(), a.to()));
				if (to <= v.currentBet()) {
					yield callOrCheck(v);
				}
				yield to >= v.maxRaiseTo() ? Action.allIn() : Action.raiseTo(to);
			}
			case CALL -> v.toCall() > 0 ? a : Action.check();
			case CHECK -> v.canCheck() ? a : Action.fold();
			case FOLD -> v.canCheck() ? Action.check() : a;
		};
	}

	/**
	 * The bot's action. {@code equity} is the Monte-Carlo result (needed for Regular/Shark after the
	 * flop; ignored otherwise).
	 */
	public static Action decide(Tier tier, View v, double equity, PokerRng rng) {
		Action a;
		if (tier == Tier.FISH) {
			a = fish(v, rng);
		} else if (v.street() == Hand.Street.PREFLOP) {
			a = preflopSolid(v, rng, tier == Tier.SHARK);
		} else if (tier == Tier.REGULAR) {
			a = regularPost(v, equity, rng);
		} else {
			a = sharkPost(v, equity, rng);
		}
		if (tier == Tier.REGULAR && rng.nextDouble() < 0.1) {
			a = lower(v, a);
		}
		return legalize(v, a);
	}

	/** Convenience: full decision incl. the Monte-Carlo run for the bot to act in {@code h}. */
	public static Action decide(Hand h, Tier tier, Map<String, Double> vpip, int regularSamples, int sharkSamples, PokerRng rng) {
		int i = h.toAct();
		View v = view(h, i, vpip);
		int samples = samplesFor(tier, h.street(), regularSamples, sharkSamples);
		double e = 0;
		if (samples > 0) {
			int[] ranges = tier == Tier.SHARK ? opponentRanges(h, i) : null;
			e = Equity.equity(v.hole(), v.board(), v.opponents(), samples, ranges, rng);
		}
		return decide(tier, v, e, rng);
	}
}

File: bedrock/src/games/poker/logic/bots.ts
/**
 * Poker bots: Fish / Regular / Shark (GAME_DESIGN §7.4). PURE.
 * A bot sees only its own cards and public information (no collusion). The game builds a
 * BotView with `botView()`, runs the Monte-Carlo equity (spread over ticks) when
 * `needsEquity()` says so, then calls `decideBot()`.
 */
import { type Rng, weightedPick } from '../../../core/logic/rng';
import { type PCard, rankOf, suitOf } from './cards';
import { type Action, type HandState, type Street, legal, livePlayers, potTotal } from './engine';
import { TOP40_CHEN, chen } from './equity';
import { categoryOf, evaluate } from './evaluator';

export const BOT_TIERS = ['fish', 'regular', 'shark'] as const;
export type BotTier = (typeof BOT_TIERS)[number];

/** Fixed, non-localized bot names (GAME_DESIGN §7.4). */
export const BOT_NAMES = [
  'Lucky Steve',
  'Grandpa Pavel',
  'Creeper42',
  'Mr. Blocksworth',
  'Diamond Dave',
  'Aunt Zoya',
  'Redstone Rick',
  'Nether Nick',
  'Emerald Emma',
  'Sir Oinksalot',
  'Baba Valya',
  'Enderman Ed',
] as const;

export type Position = 'early' | 'middle' | 'late' | 'sb' | 'bb';

export interface BotView {
  hole: PCard[];
  board: PCard[];
  street: Street;
  /** all chips in the middle (pots + current street bets) */
  pot: number;
  toCall: number;
  stack: number;
  bet: number;
  currentBet: number;
  bb: number;
  canCheck: boolean;
  canRaise: boolean;
  minRaiseTo: number;
  maxRaiseTo: number;
  position: Position;
  /** preflop: players who just called the big blind */
  limpers: number;
  /** preflop: someone raised above the big blind */
  facingRaise: boolean;
  /** this bot made the last preflop raise */
  preflopRaiser: boolean;
  /** opponents still in the hand */
  opponents: number;
  /** highest VPIP (0..1) among live human opponents (sharks adapt), if known */
  loosestHumanVpip?: number;
}

/** Preflop position of player i (early = first 2 to act, late = cutoff/button). */
export function positionOf(s: HandState, i: number): Position {
  const n = s.players.length;
  if (n === 2) return i === s.button ? 'sb' : 'bb';
  if (i === s.sbIndex) return 'sb';
  if (i === s.bbIndex) return 'bb';
  if (i === s.button) return 'late';
  const first = (s.bbIndex + 1) % n;
  const order = (i - first + n) % n;
  if (order < 2) return 'early';
  if ((i + 1) % n === s.button) return 'late';
  return 'middle';
}

/** Public view for the bot in seat i. `vpip` maps player id -> VPIP of humans (0..1). */
export function botView(s: HandState, i: number, vpip?: ReadonlyMap<string, number>): BotView {
  const p = s.players[i]!;
  const l = legal(s, i);
  const liveOpp = livePlayers(s).filter((j) => j !== i);
  let loosest: number | undefined;
  if (vpip) {
    for (const j of liveOpp) {
      const q = s.players[j]!;
      const v = q.human ? vpip.get(q.id) : undefined;
      if (v !== undefined) loosest = Math.max(loosest ?? 0, v);
    }
  }
  return {
    hole: p.hole.slice(),
    board: s.board.slice(),
    street: s.street,
    pot: potTotal(s),
    toCall: l.toCall,
    stack: p.stack,
    bet: p.bet,
    currentBet: s.currentBet,
    bb: s.bb,
    canCheck: l.canCheck,
    canRaise: l.canRaise,
    minRaiseTo: l.minRaiseTo,
    maxRaiseTo: l.maxRaiseTo,
    position: positionOf(s, i),
    limpers: s.street === 'preflop' && s.currentBet <= s.bb ? s.players.filter((q, j) => j !== i && q.vpip && !q.pfr).length : 0,
    facingRaise: s.street === 'preflop' && s.currentBet > s.bb,
    preflopRaiser: s.preflopRaiser === i,
    opponents: liveOpp.length,
    loosestHumanVpip: loosest,
  };
}

/**
 * Opponent ranges for the shark's equity: live opponents who raised preflop are assumed to
 * hold a top-40 % Chen hand.
 */
export function opponentRanges(s: HandState, i: number): (number | undefined)[] {
  return livePlayers(s)
    .filter((j) => j !== i)
    .map((j) => (s.players[j]!.pfr ? TOP40_CHEN() : undefined));
}

/** Monte-Carlo samples a tier needs on this street (0 = no simulation). */
export function samplesFor(tier: BotTier, street: Street, cfg: { regularSamples: number; sharkSamples: number }): number {
  if (tier === 'fish' || street === 'preflop') return 0;
  return tier === 'regular' ? cfg.regularSamples : cfg.sharkSamples;
}

/** Pick a tier from Fish/Regular/Shark percentages (normalized; all zero = Regular). */
export function pickTier(rng: Rng, mix: readonly number[]): BotTier {
  const w = BOT_TIERS.map((t, k) => [t, Math.max(0, Number(mix[k]) || 0)] as const);
  if (!w.some(([, x]) => x > 0)) return 'regular';
  return weightedPick(rng, w);
}

/** Total street bet for a bet/raise of `fraction` of the pot. */
export function sizeTo(v: BotView, fraction: number): number {
  if (v.currentBet === 0) return Math.max(v.bb, Math.round(v.pot * fraction));
  return Math.round(v.currentBet + fraction * (v.pot + v.toCall));
}

const raise = (to: number): Action => ({ type: 'raise', to: Math.floor(to) });
const passive = (v: BotView): Action => (v.canCheck ? { type: 'check' } : { type: 'fold' });
const callOrCheck = (v: BotView): Action => (v.toCall > 0 ? { type: 'call' } : { type: 'check' });
const potOdds = (v: BotView): number => (v.toCall > 0 ? v.toCall / (v.pot + v.toCall) : 0);

/** Made-hand category that uses at least one hole card (0 = nothing beyond the board). */
export function madeCategory(hole: readonly PCard[], board: readonly PCard[]): number {
  if (board.length < 3) return 0;
  const all = categoryOf(evaluate([...hole, ...board]));
  return all > boardCategory(board) ? all : 0;
}

function boardCategory(board: readonly PCard[]): number {
  if (board.length >= 5) return categoryOf(evaluate(board.slice(0, 7)));
  const counts = new Map<number, number>();
  for (const c of board) counts.set(rankOf(c), (counts.get(rankOf(c)) ?? 0) + 1);
  const n = [...counts.values()].sort((a, b) => b - a);
  if (n[0] === 4) return 7;
  if (n[0] === 3) return n[1] === 2 ? 6 : 3;
  if (n[0] === 2) return n[1] === 2 ? 2 : 1;
  return 0;
}

const isPocketPair = (h: readonly PCard[]): boolean => rankOf(h[0]!) === rankOf(h[1]!);
const isSuited = (h: readonly PCard[]): boolean => suitOf(h[0]!) === suitOf(h[1]!);
const isConnected = (h: readonly PCard[]): boolean => Math.abs(rankOf(h[0]!) - rankOf(h[1]!)) === 1;
const hasAce = (h: readonly PCard[]): boolean => rankOf(h[0]!) === 14 || rankOf(h[1]!) === 14;

// ---- Fish ---------------------------------------------------------------------------------

function fish(v: BotView, rng: Rng): Action {
  const jitter = 0.9 + rng.next() * 0.2;
  if (v.street === 'preflop') {
    const c = chen(v.hole[0]!, v.hole[1]!);
    if (c >= 12 && v.currentBet < 3 * v.bb) return raise(3 * v.bb * jitter);
    if (v.facingRaise) {
      if (isPocketPair(v.hole) && v.currentBet <= 10 * v.bb) return callOrCheck(v);
      if (c >= 6 && v.toCall <= 0.2 * (v.stack + v.bet)) return callOrCheck(v);
      return passive(v);
    }
    if (c >= 4) return callOrCheck(v);
    return passive(v);
  }
  const made = madeCategory(v.hole, v.board);
  if (v.toCall > 0) return made >= 1 && v.toCall <= v.pot ? { type: 'call' } : { type: 'fold' };
  if (made >= 2) return raise(sizeTo(v, 0.5 * jitter));
  if (made === 0 && rng.next() < 0.05) return raise(sizeTo(v, 0.5 * jitter));
  return { type: 'check' };
}

// ---- Regular / Shark preflop ---------------------------------------------------------------

function openThreshold(pos: Position): number {
  switch (pos) {
    case 'early':
      return 8;
    case 'middle':
    case 'bb':
      return 7;
    case 'late':
    case 'sb':
      return 6;
  }
}

function preflopSolid(v: BotView, rng: Rng, shark: boolean): Action {
  const c = chen(v.hole[0]!, v.hole[1]!);
  const d = shark ? 1 : 0;
  const widen = shark && (v.loosestHumanVpip ?? 0) > 0.4 ? 1 : 0;
  if (!v.facingRaise) {
    if (c >= openThreshold(v.position) - d) return raise(3 * v.bb + v.limpers * v.bb);
    return passive(v);
  }
  if (c >= 11 - d) return raise(3 * v.currentBet);
  if (shark && isSuited(v.hole) && (isConnected(v.hole) || hasAce(v.hole)) && rng.next() < 0.08) return raise(3 * v.currentBet);
  if (c >= 9 - d - widen) return callOrCheck(v);
  return passive(v);
}

// ---- Regular / Shark postflop --------------------------------------------------------------

function regularPost(v: BotView, e: number, rng: Rng): Action {
  if (e >= 0.65) return raise(sizeTo(v, 0.66));
  if (v.toCall > 0) return e >= potOdds(v) + 0.05 ? { type: 'call' } : { type: 'fold' };
  if (v.street === 'flop' && v.preflopRaiser && rng.next() < 0.3) return raise(sizeTo(v, 0.5));
  return { type: 'check' };
}

function sharkPost(v: BotView, e: number, rng: Rng): Action {
  const size = () => sizeTo(v, 0.5 + rng.next() * 0.25);
  if (e >= 0.85) {
    const r = rng.next();
    if (r < 0.2) return { type: 'allin' };
    if (r < 0.35) return callOrCheck(v);
    return raise(sizeTo(v, 0.75));
  }
  if (e >= 0.6) return v.toCall > 0 ? { type: 'call' } : raise(size());
  if (v.street === 'flop' && e >= 0.3 && rng.next() < 0.4) return raise(size());
  if (v.toCall > 0) return e >= potOdds(v) ? { type: 'call' } : { type: 'fold' };
  return { type: 'check' };
}

/** One step less aggressive: raise -> call/check, call -> fold/check. */
function lower(v: BotView, a: Action): Action {
  if (a.type === 'raise' || a.type === 'allin') return callOrCheck(v);
  if (a.type === 'call') return passive(v);
  return a;
}

/** Make an action legal for the view (clamp sizes, raise -> call when raising is closed). */
export function legalize(v: BotView, a: Action): Action {
  switch (a.type) {
    case 'raise':
    case 'allin': {
      if (!v.canRaise) return callOrCheck(v);
      if (a.type === 'allin') return a;
      const to = Math.max(v.minRaiseTo, Math.min(v.maxRaiseTo, Math.floor(a.to)));
      if (to <= v.currentBet) return callOrCheck(v);
      return to >= v.maxRaiseTo ? { type: 'allin' } : { type: 'raise', to };
    }
    case 'call':
      return v.toCall > 0 ? a : { type: 'check' };
    case 'check':
      return v.canCheck ? a : { type: 'fold' };
    case 'fold':
      return v.canCheck ? { type: 'check' } : a;
  }
}

/**
 * The bot's action. `equity` is the Monte-Carlo result (required for Regular/Shark after the
 * flop; pass undefined otherwise).
 */
export function decideBot(tier: BotTier, v: BotView, equity: number | undefined, rng: Rng): Action {
  let a: Action;
  if (tier === 'fish') a = fish(v, rng);
  else if (v.street === 'preflop') a = preflopSolid(v, rng, tier === 'shark');
  else if (tier === 'regular') a = regularPost(v, equity ?? 0, rng);
  else a = sharkPost(v, equity ?? 0, rng);
  if (tier === 'regular' && rng.next() < 0.1) a = lower(v, a);
  return legalize(v, a);
}
